storage/src/main/java/tech/pegasys/teku/storage/client/StoreTransactionUpdates.java

```diff
@@ -0,0 +1,251 @@
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.datastructures.blocks.BeaconBlock;
+import tech.pegasys.teku.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.datastructures.blocks.SignedBlockAndState;
+import tech.pegasys.teku.datastructures.state.BeaconState;
+import tech.pegasys.teku.datastructures.state.Checkpoint;
+import tech.pegasys.teku.datastructures.state.CheckpointAndBlock;
+import tech.pegasys.teku.storage.client.Store.Transaction;
+import tech.pegasys.teku.storage.events.StorageUpdate;
+
+class StoreTransactionUpdates {
+  private final Store.Transaction tx;
+
+  private final Map<Bytes32, SignedBeaconBlock> hotBlocks;
+  private final Map<Bytes32, SignedBlockAndState> finalizedChainData;
+  private final Map<UnsignedLong, Set<Bytes32>> prunedHotBlockRoots;
+  private final Map<Checkpoint, BeaconState> checkpointStates;
+  private final Set<Checkpoint> staleCheckpointStates;
+  private final Optional<CheckpointAndBlock> newFinalizedCheckpoint;
+  private final Map<Bytes32, BeaconState> hotStates;
+
+  private StoreTransactionUpdates(
+      final Transaction tx,
+      final Map<Bytes32, SignedBeaconBlock> hotBlocks,
+      final Map<Bytes32, BeaconState> hotStates,
+      final Map<Bytes32, SignedBlockAndState> finalizedChainData,
+      final Map<UnsignedLong, Set<Bytes32>> prunedHotBlockRoots,
+      final Map<Checkpoint, BeaconState> checkpointStates,
+      final Set<Checkpoint> staleCheckpointStates,
+      final Optional<CheckpointAndBlock> newFinalizedCheckpoint) {
+    this.tx = tx;
+    this.hotBlocks = hotBlocks;
+    this.hotStates = hotStates;
+    this.finalizedChainData = finalizedChainData;
+    this.prunedHotBlockRoots = prunedHotBlockRoots;
+    this.checkpointStates = checkpointStates;
+    this.staleCheckpointStates = staleCheckpointStates;
+    this.newFinalizedCheckpoint = newFinalizedCheckpoint;
+  }
+
+  public static StoreTransactionUpdates calculate(final Store baseStore, final Transaction tx) {
+    // Save copy of tx data that may be pruned
+    final Map<Bytes32, SignedBeaconBlock> hotBlocks = new HashMap<>(tx.blocks);
+    final Map<Bytes32, BeaconState> hotStates = new HashMap<>(tx.block_states);
+    final Map<Checkpoint, BeaconState> checkpointStates = new HashMap<>(tx.checkpoint_states);
+
+    // If a new checkpoint has been finalized, calculated what to finalize and what to prune
+    final CheckpointAndBlock prevFinalizedCheckpoint = baseStore.getFinalizedCheckpointAndBlock();
+    final Optional<CheckpointAndBlock> newFinalizedCheckpoint =
+        Optional.of(tx.getFinalizedCheckpointAndBlock())
+            .filter(c -> c.getEpoch().compareTo(prevFinalizedCheckpoint.getEpoch()) > 0);
```

Comment:
Looks like it would be easier to not use an `Optional` here.  Just something like:
```
final CheckpointAndBlock newFinalizedCheckpoint = tx.getFinalizedCheckpointAndBlock();
...
if (newFinalizedCheckpoint.getEpoch().compareTo(prevFinalizedCheckpoint.getEpoch()) > 0) {
    ... calculate stuff to prune ...
} else {
  .. the empty map stuff ...
}
```

Although I guess you need the optional to pass into the `StoreTransactionUpdates` so happy to go either way.